diff --git i/agent/xds/config.go w/agent/xds/config.go
index 020cefeb5..e672b97ce 100644
--- i/agent/xds/config.go
+++ w/agent/xds/config.go
@@ -38,6 +38,12 @@ type ProxyConfig struct {
 	// set.
 	LocalConnectTimeoutMs int `mapstructure:"local_connect_timeout_ms"`
 
+	// LocalIdleTimeoutMs is the number of milliseconds a request's stream to the
+	// local app instance may be idle. If not set, no value is set, Envoy defaults
+	// are respected, and an Envoy stream_idle_timeout (5m) will apply. If set,
+	// this LocalIdleTimeoutMs value will override the Envoy stream_idle_timeout.
+	LocalIdleTimeoutMs *int `mapstructure:"local_idle_timeout_ms"`
+
 	// LocalRequestTimeoutMs is the number of milliseconds to timeout HTTP requests
 	// to the local app instance. If not set, no value is set, Envoy defaults are
 	// respected (15s)
diff --git i/agent/xds/listeners.go w/agent/xds/listeners.go
index 319161ff4..b0b6a45df 100644
--- i/agent/xds/listeners.go
+++ w/agent/xds/listeners.go
@@ -837,6 +837,7 @@ func (s *ResourceGenerator) makeInboundListener(cfgSnap *proxycfg.ConfigSnapshot
 		filterName:       name,
 		routeName:        name,
 		cluster:          LocalAppClusterName,
+		idleTimeoutMs:    cfg.LocalIdleTimeoutMs,
 		requestTimeoutMs: cfg.LocalRequestTimeoutMs,
 	}
 	if useHTTPFilter {
@@ -1347,6 +1348,7 @@ type listenerFilterOpts struct {
 	cluster          string
 	statPrefix       string
 	routePath        string
+	idleTimeoutMs    *int
 	requestTimeoutMs *int
 	ingressGateway   bool
 	httpAuthzFilter  *envoy_http_v3.HttpFilter
@@ -1455,6 +1457,11 @@ func makeHTTPFilter(opts listenerFilterOpts) (*envoy_listener_v3.Filter, error)
 			},
 		}
 
+		if opts.idleTimeoutMs != nil {
+			r := route.GetRoute()
+			r.IdleTimeout = ptypes.DurationProto(time.Duration(*opts.idleTimeoutMs) * time.Millisecond)
+		}
+
 		if opts.requestTimeoutMs != nil {
 			r := route.GetRoute()
 			r.Timeout = ptypes.DurationProto(time.Duration(*opts.requestTimeoutMs) * time.Millisecond)
